Applications/Spire/Source/Charting/CachedChartModel.cpp

```diff
@@ -17,159 +20,127 @@ ChartValue::Type CachedChartModel::get_y_axis_type() const {
+            return std::move(data);
+          });
       }
-      return std::vector<Candlestick>(first_index, last_index);
+      on_data_loaded(std::move(result.Get()), info);
+      return load_from_cache(info);
     });
 }
 
 void CachedChartModel::on_data_loaded(const std::vector<Candlestick>& data,
-    ChartValue first, ChartValue last) {
-  for(auto& range : m_ranges) {
-    if(range.m_start <= first && range.m_end >= last) {
-      return;
-    }
-  }
-  insert_data(data);
-  update_ranges(first, last);
-}
-
-void CachedChartModel::update_ranges(ChartValue first, ChartValue last) {
-  if(m_ranges.empty()) {
-    m_ranges.push_back({first, last});
+    const continuous_interval<ChartValue>& loaded_range) {
+  if(contains(m_ranges, loaded_range)) {
     return;
   }
-  auto remove_range = [] (auto& ranges, auto& range) {
-    for(auto iter = ranges.begin(); iter != ranges.end(); ++iter) {
-      if((*iter).m_start == range.m_start && (*iter).m_end == range.m_end) {
-        ranges.erase(iter);
-        break;
+  auto first_iterator = data.begin();
+  auto last_iterator = data.end();
+  if(!data.empty()) {
+    for(auto& range : m_ranges) {
+      auto first = continuous_interval<ChartValue>::closed(
+        loaded_range.lower(), loaded_range.lower());
+      if(intersects(first, range)) {
+        first_iterator = std::upper_bound(data.begin(), data.end(),
+          range.upper(),
+          [] (const auto& value, const auto& index) {
+            return index.GetStart() >= value;
+          });
+        if(first_iterator != data.end()) {
+          ++first_iterator;
+        }
+      }
+      auto last = continuous_interval<ChartValue>::closed(
+        loaded_range.upper(), loaded_range.upper());
+      if(intersects(last, range)) {
+        last_iterator = std::lower_bound(data.begin(), data.end(),
+          range.lower(),
+          [] (const auto& index, const auto& value) {
+            return index.GetEnd() <= value;
+          });
+        if(data.size() > 1 && first_iterator < last_iterator) {
+          --last_iterator;
+        }
       }
-    }
-  };
-  auto ranges = m_ranges;
-  auto new_first = first;
-  auto new_last = last;
-  for(auto& range : m_ranges) {
-    if(range.m_end < first) {
-      continue;
-    }
-    if(range.m_start > last) {
-      break;
-    }
-    if(range.m_start <= first && range.m_end >= first) {
-      new_first = range.m_start;
-      remove_range(ranges, range);
-    } else if(range.m_start <= last && range.m_end >= last) {
-      new_last = range.m_end;
-      remove_range(ranges, range);
-    }
-    if(range.m_start >= first && range.m_end <= last) {
-      remove_range(ranges, range);
     }
   }
-  auto index = std::lower_bound(ranges.begin(), ranges.end(), first,
-    [] (const auto& index, const auto& value) {
-      return index.m_start < value;
-    });
-  ranges.insert(index, ChartRange{new_first, new_last});
-  m_ranges = ranges;
+  m_cache.store(std::move(
```

Comment:
I claim you don't need a ```move``` here, why am I right or wrong?

Reply:
You're correct because the vector is already an rvalue.

Reply:
Good good!